pallets/gear/src/lib.rs

```diff
@@ -562,39 +569,30 @@ pub mod pallet {
                     )
                 };
 
-                core_processor::handle_journal(journal.clone(), &mut ext_manager);
+                // TODO: Check whether we charge gas fee for submitting code after #646
+                for note in journal.into_iter() {
+                    core_processor::handle_journal(vec![note.clone()], &mut ext_manager);
```

Comment:
why we still need to handle every journal on this? looks like only
- GasBurned
- WakeMessage
- SendDispatch
- WaitDispatch

spend gas

Reply:
We should handle all journal to apply changes and provide messages.

Lets go through the `get_gas_spent` idea:
It's an RPC call to get the amount of gas needed to success execution.
It works like this:
- We set for the initial message maximum of available gas limit: `T::BlockGasLimit::get()`
- We clean current message queue (It has no effect for storage in RPC-call processing due to overlay, but applies changes on storage inside test, be careful with it)
- We insert initial message into the queue
- We run our queue processing till the moment where there is no messages left
- On each iteration we parse ALL created journal (its notes) to apply proper storage changes, emulating real execution
- To don't be chained with concrete notes (avoiding future collisions, cause we may forget about `get_gas_spent`), we go through all of them in custom parsing - after each note processed we query current gas for initial message to find maximal diff from the initial one - this would be the resulting value we should return to user.

`get_gas_burned` - works the same, but we don't calculate minimal amount of gas needed for success execution (that maximal diff) - we just add up all gas burned to get the diff of the user balance after all executions, cause only this gas won't be returned on consume of the user message.

This stuff needs optimisation and code beautifying, but atm this is solution for issues we ran into.